### PR TITLE
[WIP] docs(cb-http-interception): add Http interception cookbook

### PR DESCRIPTION
**Pull request by wardbell**
**Work in Progress - do not merge**

Todo:
- [ ] Confirm this is a good approach with DevRel
- [ ] Write prose

### CODE REVIEW

**Comment by bennadel** [comment]
@wardbell so I've been staring at this code for a while and I think I'm having some issues reconciling the NG2 approach here with the existing NG1 approach. If we look at the NG1 code, there are a few differences:

1. Request / Response interceptors are different. Or at least, they can be defined as separate properties on the interceptor, which means the user-code itself doesn't need to differentiate between the incoming/outgoing when processing actions.
2. The interceptors are setup during the "config" phase.

Now, I know we don't really have a "config" phase in NG2, so maybe putting the setup in the root-component is the way to go. After all, it is the root of the application. 

The other thing that strikes me is the use of `.subscribe()` as a means to *alter* an action. Something about that feels strange. When I think about "subscribe" in other aspects, such as RxJS streams, it makes me think that I have access to the "result" of something. In this case, however, the subscribe is to the in-process workflow. Just feels a little odd.

I'm still noodling on this, but just some thoughts I had.

**Comment by bennadel** [comment]
Going back to the config phase, I almost wish we had some sort of DI token for these things. Like:

```js
{
    provide: HTTP_INTERCEPTORS,
    multi: true,
    useClass: MyHttpInterceptor
}
```

... where that class could implement something more akin to the NG1 API:

```js
class MyHttpInterceptor implements HttpIncterceptor {
    public function request( config ) {}
    public function requestError( rejection ) {}
    public function response( config ) {}
    public function responseError( rejection ) {}
}
```

Then, maybe this class could also add another "Subject" for the root-node to subscribe to for the logging.

**Comment by wardbell** [comment]
@bennadel Ok ... I agree that my previous approach was unintuitive.  So I've completely rewritten it. What do you think now?

**Comment by wardbell** [comment]
Angular is adding HTTP interception; no point in this one.

File: public/docs/_examples/cb-http-interception/ts/app/http-interceptor.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  addRequestInterceptor(fn: interceptorFn) {
+    this.requestInterceptors.push(fn);
+  }
+
+  addResponseInterceptor(fn: interceptorFn) {
+    this.responseInterceptors.push(fn);
+  }
+
+  addResponseErrorInterceptor(fn: interceptorFn) {
+    this.responseInterceptors.push(fn);
```
Comment (d-nation) [inline public/docs/_examples/cb-http-interception/ts/app/http-interceptor.ts]: For anyone who is using this until HTTP Interceptors land... this line should read:
`this.responseErrorInterceptors.push(fn);`

**Pull request closed without merging**